### PR TITLE
feat(testworkflows): add support for pausing steps

### PR DESCRIPTION
**Pull request by rangoo94**
## Pull request description 

* add `paused` property for initial pause in the step (https://github.com/kubeshop/testkube-operator/pull/245)
* add a control server to the Init Process, that allows pausing/resuming the step
* add API endpoints to call the control server
* pause the processes at any time
   * when paused initially, the command won't start at all until resumed
   * when paused during the execution, `SIGSTOP` signal will be sent to all children processes recursively, when resumed - `SIGCONT`
      * this allows to pause any kind of process, but may cause losing TCP connections (due to inactivity)
* fix: avoid empty lines in log for consecutive outputs

## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [x] tested locally
- [x] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: pkg/tcl/testworkflowstcl/testworkflowprocessor/initprocess.go [hunk 1]
```diff
@@ -192,11 +196,20 @@ func (p *initProcess) AddRetryPolicy(policy testworkflowsv1.RetryPolicy, ref str
 	return p
 }
 
+func (p *initProcess) SetPaused(paused string) *initProcess {
+	if paused == "null" || paused == "false" || paused == "never" {
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply pkg/tcl/testworkflowstcl/testworkflowprocessor/initprocess.go:200]: Not really, these are just common shortcuts that may happen there.

I can though change it to run that expression and change to `""` if the expression result is falsy 👍

File: pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go [hunk 2]
```diff
@@ -118,6 +123,40 @@ func (c *controller) Cleanup(ctx context.Context) error {
...
+func (c *controller) command(ctx context.Context, name string, body io.Reader) error {
+	// TODO: add waiting for the started container + retries?
+	v := <-c.pod.Any(ctx)
+	if v.Error != nil {
+		return v.Error
+	}
+	if v.Value.Status.PodIP == "" {
+		return errors.New("there is no IP assigned to this pod")
+	}
+	r, err := http.NewRequest("POST", fmt.Sprintf("http://%s:%d/%s", v.Value.Status.PodIP, constants2.ControlServerPort, name), body)
```
Comment (vsukhin) [inline pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go:135]: http.MethodPost

File: pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go [hunk 3]
```diff
@@ -322,7 +361,15 @@ func (c *controller) Watch(parentCtx context.Context) Watcher[Notification] {
 					continue
 				}
 				if v.Value.Hint != nil {
-					if v.Value.Hint.Name == "start" && v.Value.Hint.Ref == container.Name {
+					if v.Value.Hint.Name == "pause" {
```
Comment (vsukhin) [inline pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go:364]: looks like const

File: pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go [hunk 4]
```diff
@@ -322,7 +361,15 @@ func (c *controller) Watch(parentCtx context.Context) Watcher[Notification] {
...
+						start, _ := time.Parse(constants2.PreciseTimeFormat, v.Value.Hint.Value.(string))
+						result.PauseStart(sig, c.scheduledAt, v.Value.Hint.Ref, start)
+						w.SendValue(Notification{Result: result.Clone()})
+					} else if v.Value.Hint.Name == "resume" {
+						end, _ := time.Parse(constants2.PreciseTimeFormat, v.Value.Hint.Value.(string))
```
Comment (vsukhin) [inline pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go:369]: shouldn't we log parsing error?

**Review by vsukhin — APPROVED** [review]

**Pull request merged**